### PR TITLE
Address ISLANDORA-2124; Update PDF.js version, add Solr alter hooks t…

### PR DESCRIPTION
**Pull request by nhart**
…o pass page and search term to viewer.

**JIRA Ticket**: https://jira.duraspace.org/browse/ISLANDORA-2124

https://groups.google.com/forum/#!topic/islandora/3EiX1w4xsko

# What does this Pull Request do?

Adds the necessary Solr hooks so the book page and Solr search term are passed to the PDF.js viewer. This allows the viewer to highlight the term and brings the user to appropriate page of the document.

# What's new?
Adds "page" and "search" params to the Solr result url to be used when calling the PDf.js viewer.html file.

# How should this be tested?
See Jira ticket for details.

# Additional Notes:
Requires updated PDF.js viewer

# Interested parties
@Islandora/7-x-1-x-committers

### CODE REVIEW

**Comment by DiegoPino** [comment]
Oh, this is great @nhart. Testing

File: islandora_pdfjs.module [hunk 1]
```diff
@@ -78,3 +78,78 @@ function islandora_pdfjs_viewer_callback(array $params, $fedora_object = NULL) {
...
   }
 }
+
+/**
+ * Implements hook_CMODEL_PID_islandora_solr_object_result_alter().
+ *
+ * Add page viewing fragment and search term to show all search results within
+ * book on page load.
+ */
+function islandora_pdfjs_islandora_bookCModel_islandora_solr_object_result_alter(&$search_results, $query_processor) {
```
Comment (DiegoPino) [inline islandora_pdfjs.module]: A bit confused here. Why the BookCmodel and/or not the PDF CMODEL?

Comment (nhart) [reply islandora_pdfjs.module]: Hum. good point.. I guess we could do PDF too and a few others maybe.

Comment (nhart) [reply islandora_pdfjs.module]: Was mainly copying IAV functionality to pdf.js. Will update pr.

**Comment by nhart** [comment]
There are also some changes that can be added to webViewerInitialized function in the viewer.js file that make this PR nice.. Not sure how we feel about forking the PDF.js library though.

var hash = document.location.hash.substring(1);
var hashParams = (0, _ui_utils.parseQueryString)(hash);
if ('search' in hashParams) {
    PDFViewerApplication.findBar.open();
    PDFViewerApplication.findBar.findField.value = hashParams['search'];
    PDFViewerApplication.findBar.caseSensitive.checked = true;
    PDFViewerApplication.findBar.highlightAll.checked = true;
    PDFViewerApplication.findBar.findNextButton.click();
}

**Comment by DiegoPino** [comment]
@nhart can the library be extended/overriden via an external file/class? Or you mean actual cloning-modifying?

**Comment by nhart** [comment]
From what I've read thus far, we'd need to fork the viewer.html file and add our own .js or fork both viewer.html and viewer.js and add our changes directly to viewer.js

**Comment by DiegoPino** [comment]
Ok, reading https://github.com/mozilla/pdf.js#learning , seem like cloning/editing is the only way...
Want to discuss this on our next committer call? I would personally say maybe dealing with PHP part on this comm for sure, then a second one to improve the JS piece. Could be easier to digest/test and get general consensus?

File: islandora_pdfjs.module (hunk 1 above)
Comment (nhart) [reply islandora_pdfjs.module]: So.. Not sure how far to run with this. basically anything that matches https://github.com/Islandora/islandora_pdfjs/blob/7.x/islandora_pdfjs.module#L42-L47 could potentially use pdf.js as a viewer. 

There's no islandora_get_viewer_id_by_cmodel or viewer id registry function so we'd need to store vars islandora_book_viewers, islandora_book_page_viewers, islandora_newspaper_issue_viewers, etc. to be able to only add the url params when the particular content model was using pdf.js

thoughts?

**Comment by DiegoPino** [comment]
@nhart i'm taking the afternoon off (tired/sick) so won't be able to bring this to this week's committer's call (Nov 30 2017). Not sure who is chairing (i now at least it is not me) Let me know if you could get consensus/ideas on using our own .js and/or HTML file and how to proceed. Will be watching this pull to make sure it gets reviewed and merged.

File: islandora_pdfjs.module [hunk 2]
```diff
@@ -144,12 +153,20 @@ function islandora_pdfjs_islandora_pageCModel_islandora_solr_object_result_alter
...
-      // highlighting field is not the one being search on, this makes sense?
-      if ($query_processor->solrDefType == 'dismax' || $query_processor->solrDefType == 'edismax') {
-        $search_results['object_url_params']['search'] = rawurlencode($query_processor->solrQuery);
-      }
     }
   }
 }
+
+/*
+ * Add search term to result url as a param to show all search results within the PDF.js viewer.
```
Comment (DiegoPino) [inline islandora_pdfjs.module]: This line is failing to pass coding standards. Over 80 chars.

**Comment by DiegoPino** [comment]
@nhart any chance you can address the DCS issue? Would love to see this in the new release. let me know.

**Comment by nhart** [comment]
Sorry, completely dropped the ball on this. I've issued a commit to fit the doc.

**Comment by nhart** [comment]
any @Islandora/7-x-1-x-committers want to review, test, and merge?

**Comment by DiegoPino** [comment]
@nhart will do during the week

File: islandora_pdfjs.module [hunk 3]
```diff
@@ -78,3 +78,95 @@ function islandora_pdfjs_viewer_callback(array $params, $fedora_object = NULL) {
...
+  }
+}
+
+/**
+ * Implements hook_CMODEL_PID_islandora_solr_object_result_alter().
+ */
+function islandora_pdfjs_islandora_newspaperIssueCModel_islandora_solr_object_result_alter(&$search_results, $query_processor) {
+  module_load_include('inc', 'islandora', 'includes/solution_packs');
+  $viewer_id = islandora_get_viewer_id('islandora_newspaper_issue_viewers');
+  if ($viewer_id != 'islandora_pdfjs') {
```
Comment (bencomp) [inline islandora_pdfjs.module]: Should you check for equality or inequality here? Above you used `==`, here you use `!=`.

Comment (DiegoPino) [reply islandora_pdfjs.module]: @nhart @bencomp makes a valid point here. Should the viewer_id here be 'islandora_pdfjs' to trigger  islandora_pdfjs_add_solr_search_param?

**Review by DiegoPino — CHANGES_REQUESTED** [review]

File: islandora_pdfjs.module (hunk 3 above)
Comment (nhart) [reply islandora_pdfjs.module]: Looks like a copy and paste error.

**Comment by nhart** [comment]
Anything else on this @bencomp @DiegoPino ?

**Review by DiegoPino — APPROVED** [review]

**Comment by DiegoPino** [comment]
Looks good to me @nhart. 24 hours and merging. Can you let me know if I forget? Not too many merging and reviewing this week

**Comment by DonRichards** [comment]
@DiegoPino bump

**Comment by nhart** [comment]
thank you!

**Pull request merged**